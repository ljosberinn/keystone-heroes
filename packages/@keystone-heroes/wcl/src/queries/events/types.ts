export type BaseEvent<T extends Record<string, unknown>> = T & {
  timestamp: number;
  // pin: "0" but irrelevant
};

type EncounterStartEvent = BaseEvent<{
  type: "encounterstart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
  level: number;
  affixes: number[];
}>;

type CombatantInfoEvent = BaseEvent<{
  type: "combatantinfo";
  sourceID: number;
  gear: [];
  auras: {
    source: number;
    ability: number;
    stacks: number;
    icon: string;
    name?: string;
  }[];
  expansion: string;
  faction: number;
  specID: number;
  covenantID: number;
  soulbindID: number;
  strength: number;
  agility: number;
  stamina: number;
  intellect: number;
  dodge: number;
  parry: number;
  block: number;
  armor: number;
  critMelee: number;
  critRanged: number;
  critSpell: number;
  speed: number;
  leech: number;
  hasteMelee: number;
  hasteRanged: number;
  hasteSpell: number;
  avoidance: number;
  mastery: number;
  versatilityDamageDone: number;
  versatilityHealingDone: number;
  versatilityDamageReduction: number;
  talents: { id: number; icon: string }[];
  pvpTalents: { id: number; icon: string }[];
  artifact: { traitID: number; rank: number; spellID: number; icon: string }[];
  heartOfAzeroth: {
    traitID: number;
    rank: number;
    spellID: number;
    icon: string;
  }[];
}>;

export type ApplyDebuffEvent = BaseEvent<{
  type: "applydebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  targetInstance?: number;
  targetMarker?: number;
}>;

export type ApplyDebuffStackEvent = BaseEvent<{
  type: "applydebuffstack";
  stack: number;
  abilityGameID: number;
}>;

export type ApplyBuffEvent = BaseEvent<{
  type: "applybuff";
  sourceID: number;
  sourceInstance?: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
}>;

export type RemoveBuffEvent = BaseEvent<{
  type: "removebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type CastEvent = BaseEvent<{
  type: "cast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker?: number;
}>;

export type DamageEvent = BaseEvent<{
  type: "damage";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  mitigated?: number;
  unmitigatedAmount?: number;
  tick?: boolean;
  sourceMarker?: number;
  absorbed?: number;
  overkill?: number;
  buffs?: string;
}>;

export type BeginCastEvent = BaseEvent<{
  type: "begincast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

export type RemoveDebuffEvent = BaseEvent<{
  type: "removedebuff";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type SummonEvent = BaseEvent<{
  type: "summon";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type PhaseStartEvent = BaseEvent<{
  type: "phasestart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
}>;

export type HealEvent = BaseEvent<{
  type: "heal";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  sourceMarker?: number;
  targetMarker?: number;
  tick?: boolean;
  overheal?: number;
  absorbed?: number;
}>;

type EnergizeEvent = BaseEvent<{
  type: "energize";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  resourceChange: number;
  resourceChangeType: number;
  otherResourceChange: number;
  waste: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type ApplyBuffStackEvent = BaseEvent<{
  type: "applybuffstack";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  stack: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type InterruptEvent = BaseEvent<{
  type: "interrupt";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  sourceMarker?: number;
}>;

export type AbsorbEvent = BaseEvent<{
  type: "absorbed";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  attackerID: number;
  amount: number;
  extraAbilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type RefreshDebuffEvent = BaseEvent<{
  type: "refreshdebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

type DispelEvent = BaseEvent<{
  type: "dispel";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  isBuff: boolean;
  sourceMarker: number;
  targetMarker: number;
}>;

export type DeathEvent = BaseEvent<{
  type: "death";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  // only present if NPC kills player
  killerID?: number;
  killerInstance?: number;
  killingAbilityGameID?: number;
  targetMarker?: number;
}>;

export type AnyEvent =
  | DeathEvent
  | DispelEvent
  | RefreshDebuffEvent
  | AbsorbEvent
  | InterruptEvent
  | ApplyBuffStackEvent
  | EnergizeEvent
  | HealEvent
  | PhaseStartEvent
  | SummonEvent
  | RemoveDebuffEvent
  | BeginCastEvent
  | DamageEvent
  | CastEvent
  | RemoveBuffEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
  | ApplyDebuffEvent
  | CombatantInfoEvent
  | EncounterStartEvent;

export type AllTrackedEventTypes = (
  | CastEvent
  | DeathEvent
  | AbsorbEvent
  | DamageEvent
  | HealEvent
  | InterruptEvent
  | BeginCastEvent
  | ApplyBuffEvent
  | ApplyDebuffEvent
  | ApplyDebuffStackEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[];
