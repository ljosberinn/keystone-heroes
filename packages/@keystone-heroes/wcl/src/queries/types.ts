import type { PlayableClass, Role, SpecName } from "@prisma/client";

export type Table = {
  totalTime: number;
  itemLevel: number;
  logVersion: number;
  gameVersion: number;
  composition: Composition[];
  damageDone: DamageDone[];
  healingDone: HealingDone[];
  damageTaken: DamageTaken[];
  deathEvents: DeathEvent[];
  playerDetails: PlayerDetails;
};

type Composition = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  specs: SpecMeta[];
};

type SpecMeta = {
  spec: SpecName;
  role: Role;
};

export type DamageDone = Pick<Composition, "name" | "id" | "guid"> & {
  total: number;
  icon: string;
  type: Composition["type"] | "Pet" | "NPC";
};

export type HealingDone = DamageDone;

export type DamageTaken = Pick<DamageDone, "name" | "guid" | "total"> & {
  abilityIcon: string;
  type: number;
};

type DeathEvent = Pick<DamageDone, "name" | "id" | "guid" | "type" | "icon"> & {
  deathTime: number;
  ability: Omit<DamageTaken, "total">;
};

export type PlayerDetails = {
  tanks: InDepthCharacterInformation[];
  dps: InDepthCharacterInformation[];
  healers: InDepthCharacterInformation[];
};

export type InDepthCharacterInformation = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  server: string;
  icon: string;
  specs: SpecName[];
  minItemLevel: number;
  maxItemLevel: number;
  combatantInfo: CombatantInfo;
};

type CombatantInfo = {
  stats: Stats;
  talents: Talent[];
  gear: (Item | LegendaryItem)[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  artifact?: CustomPowerSet[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  heartOfAzeroth?: SecondaryCustomPowerSet[];
  customPowerSet: CustomPowerSet[];
  secondaryCustomPowerSet: SecondaryCustomPowerSet[];
  tertiaryCustomPowerSet: unknown[];
  specIDs: number[];
  factionID: number;
  covenantID?: number;
  soulbindID?: number;
};

type Stats = {
  Speed: Stat;
  Dodge: Stat;
  Mastery: Stat;
  Stamina: Stat;
  Haste: Stat;
  Leech: Stat;
  Armor: Stat;
  Agility: Stat;
  Crit: Stat;
  "Item Level": Stat;
  Parry: Stat;
  Avoidance: Stat;
  Versatility: Stat;
  Intellect: Stat;
};

type Stat = {
  min: number;
  max: number;
};

export type Talent = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
};

type CustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export enum ItemQuality {
  POOR = 1,
  COMMON = 2,
  RARE = 3,
  SUPERIOR = 4,
  LEGENDARY = 5,
}

export type Item = {
  id: number;
  slot: number;
  quality: ItemQuality;
  icon: string;
  name?: string;
  itemLevel: number;
  bonusIDs?: number[];
  gems?: Gem[];
  permanentEnchant?: number;
  permanentEnchantName?: string;
  onUseEnchant?: number;
  onUseEnchantName?: string;
  effectID?: number;
  effectIcon?: string;
  effectName?: string;
  temporaryEnchant?: number;
  temporaryEnchantName?: string;
};

export type LegendaryItem = Item & {
  quality: ItemQuality.LEGENDARY;
  effectID: number;
  effectIcon: string;
  effectName: string;
};

type Gem = {
  id: number;
  itemLevel: number;
  icon: string;
};

type SecondaryCustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export type Conduit = SecondaryCustomPowerSet;
export type CovenantTrait = CustomPowerSet;
