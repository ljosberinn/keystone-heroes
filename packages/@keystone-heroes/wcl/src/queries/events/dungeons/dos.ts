import type { AllTrackedEventTypes, ApplyDebuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const DOS_URN = 228_626;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=2&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20228626
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 34007, endTime: 1759795, filterExpression: "type = \"applydebuff\" and target.type = \"npc\" and ability.id = 228626") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "npc" and ability.id = ${DOS_URN}`,
];

const isDosUrnEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: DOS_URN,
});

export const getDOSEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return allEvents.filter(isDosUrnEvent);
};
