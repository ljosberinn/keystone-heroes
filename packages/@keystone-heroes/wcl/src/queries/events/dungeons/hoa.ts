import type { AllTrackedEventTypes, CastEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const HOA_GARGOYLE = 342_171 as const;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=4&type=summary&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22cast%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20342171&view=events
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [4]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 5987472, endTime: 7736379, filterExpression: "type = \"cast\" and source.type = \"player\" and ability.id = 342171") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "cast" and source.type = "player" and ability.id = ${HOA_GARGOYLE}`,
];

const isHoaGargoyleEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: HOA_GARGOYLE,
});

export const getHOAEvents = (
  allEvents: AllTrackedEventTypes[]
): CastEvent[] => {
  return allEvents.filter(isHoaGargoyleEvent);
};
