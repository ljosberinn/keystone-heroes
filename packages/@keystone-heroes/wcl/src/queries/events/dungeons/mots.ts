import type { ApplyDebuffEvent, AllTrackedEventTypes } from "../types";
import { createChunkByThresholdReducer, createIsSpecificEvent } from "../utils";

export const ENVELOPMENT_OF_MISTS = 323_881 as const;

/**
 * @see https://www.warcraftlogs.com/reports/aT8ZhP1HRfJANKCj#fight=18&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20323881
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "aT8ZhP1HRfJANKCj") {
 *       fights(fightIDs: [18]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 34782210, endTime: 36664869, filterExpression: "type = \"applydebuff\" and target.type = \"player\" and ability.id = 323881") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "player" and ability.id = ${ENVELOPMENT_OF_MISTS}`,
];

const isMistsMazeDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: ENVELOPMENT_OF_MISTS,
});

// failing the mists maze applies a debuff for 4 seconds
// hence its impossible to trigger it 2x within that timeframe
const debuffReducer = createChunkByThresholdReducer(4 * 1000);

export const getMOTSEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return (
    allEvents
      .filter(isMistsMazeDebuffEvent)
      .reduce<ApplyDebuffEvent[][]>(debuffReducer, [])
      // pick only the first event of each chunk,
      // indicating when the maze was failed
      .flatMap((chunk) => chunk[0])
  );
};
