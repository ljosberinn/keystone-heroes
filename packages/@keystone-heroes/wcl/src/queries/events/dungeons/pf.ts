import type {
  AllTrackedEventTypes,
  AnyEvent,
  ApplyBuffEvent,
  DamageEvent,
  DeathEvent,
} from "../types";
import {
  createChunkByThresholdReducer,
  createIsSpecificEvent,
  reduceEventsByPlayer,
} from "../utils";

const RED_BUFF = {
  unit: 164_705,
  aura: 340_225,
  // buff: 340_227,
} as const;

const PURPLE_BUFF = {
  unit: 164_707,
  aura: 340_271,
  // buff: 340_273,
} as const;

const GREEN_BUFF = {
  unit: 163_891,
  aura: 340_210,
  // buff: 340_211,
} as const;

export const PF = {
  PLAGUE_BOMB: 328_501,
  RIGGED_PLAGUEBORER: 168_878,
  GREEN_BUFF,
  RED_BUFF,
  PURPLE_BUFF,
} as const;

/**
 * @see https://www.warcraftlogs.com/reports/qGV26X4kLbRFBJDt#fight=5&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24(type%20%3D%20%22death%22%20and%20target.type%20%3D%20%22npc%22%20and%20target.id%20in%20(164705,164707,%20163891))%20or%20(type%20%3D%20%22applybuff%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20in%20(340225,%20340271,%20340210))%20or%20(type%20%3D%20%22damage%22%20and%20ability.id%20%3D%20328501)
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "qGV26X4kLbRFBJDt") {
 *       fights(fightIDs: [5]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 3958680, endTime: 6113126, filterExpression: "(type = \"death\" and target.type = \"npc\" and target.id in (164705,164707, 163891)) or (type = \"applybuff\" and target.type = \"player\" and ability.id in (340225, 340271, 340210)) or (type = \"damage\" and ability.id = 328501)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "death" and target.type = "npc" and target.id in (${RED_BUFF.unit}, ${PURPLE_BUFF.unit}, ${GREEN_BUFF.unit})`,
  `type = "applybuff" and target.type = "player" and ability.id in (${RED_BUFF.aura}, ${PURPLE_BUFF.aura}, ${GREEN_BUFF.aura})`,
  `type = "damage" and ability.id = ${PF.PLAGUE_BOMB}`,
];

const createIsPfSlimeDeathEvent =
  (actorIDSet: Set<number>) =>
  (event: AnyEvent): event is DeathEvent => {
    return (
      event.type === "death" &&
      !actorIDSet.has(event.targetID) &&
      "targetInstance" in event &&
      event.targetInstance !== undefined
    );
  };

const isPfSlimeBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: [GREEN_BUFF.aura, RED_BUFF.aura, PURPLE_BUFF.aura],
});

const isPfPlagueBombDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: PF.PLAGUE_BOMB,
});

// ApplyBuff events fire a lot for the PF Slime Auras for some reason
// Fleshcraft gets its CD reduced, but its unlikely to reduce it by 75%
const pfSlimeBuffReducer = createChunkByThresholdReducer(30 * 1000);

export const getPFEvents = (
  allEvents: AllTrackedEventTypes[],
  actorIDSet: Set<number>
): (DeathEvent | DamageEvent | ApplyBuffEvent)[] => {
  const plagueBombDamageEvents = allEvents.filter(isPfPlagueBombDamageEvent);

  const plagueBombDamageTakenEvents = reduceEventsByPlayer(
    plagueBombDamageEvents.filter((event) => actorIDSet.has(event.targetID)),
    "targetID"
  );
  const plagueBombDamageDoneEvents = plagueBombDamageEvents.filter(
    (event) => !actorIDSet.has(event.targetID)
  );

  const aggregatedPlagueBombDamageEvent =
    plagueBombDamageDoneEvents.length > 0
      ? plagueBombDamageDoneEvents.reduce((acc, event) => {
          if (acc === event) {
            return acc;
          }

          return {
            ...acc,
            amount: acc.amount + event.amount - (event.overkill ?? 0),
          };
        }, plagueBombDamageDoneEvents[0])
      : null;

  const buffEvents = allEvents.filter(isPfSlimeBuffEvent).reduce<{
    red: ApplyBuffEvent[];
    green: ApplyBuffEvent[];
    purple: ApplyBuffEvent[];
  }>(
    (acc, event) => {
      switch (event.abilityGameID) {
        case GREEN_BUFF.aura:
          return { ...acc, green: [...acc.green, event] };
        case RED_BUFF.aura:
          return { ...acc, red: [...acc.red, event] };
        case PURPLE_BUFF.aura:
          return { ...acc, purple: [...acc.purple, event] };
      }

      return acc;
    },
    {
      red: [],
      green: [],
      purple: [],
    }
  );

  return [
    ...allEvents.filter(createIsPfSlimeDeathEvent(actorIDSet)),
    ...buffEvents.green
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...buffEvents.red
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...buffEvents.purple
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...plagueBombDamageTakenEvents,
    ...(aggregatedPlagueBombDamageEvent
      ? [aggregatedPlagueBombDamageEvent]
      : []),
  ];
};
