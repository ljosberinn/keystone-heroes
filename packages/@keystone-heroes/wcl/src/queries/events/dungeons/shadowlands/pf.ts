import { EventDataType, HostilityType } from "../../../../types";
import { getEnemyNPCIDs } from "../../../report";
import type { DeathEvent, ApplyBuffEvent, RemoveBuffEvent } from "../../types";
import type { GetEventBaseParams } from "../../utils";
import { getEvents } from "../../utils";

export const PF_RED_BUFF = {
  unit: 164_705,
  aura: 340_225,
  buff: 340_227,
} as const;

export const PF_PURPLE_BUFF = {
  unit: 164_707,
  aura: 340_271,
  buff: 340_273,
} as const;

export const PF_GREEN_BUFF = {
  unit: 163_891,
  aura: 340_210,
  buff: 340_211,
} as const;

type PFSlimeKillsParams = GetEventBaseParams & {
  fightID: number;
};

export const getPFSlimeKills = async (
  params: PFSlimeKillsParams
): Promise<(DeathEvent | ApplyBuffEvent)[]> => {
  const {
    [PF_RED_BUFF.unit]: red,
    [PF_GREEN_BUFF.unit]: green,
    [PF_PURPLE_BUFF.unit]: purple,
  } = await getEnemyNPCIDs(
    {
      fightIDs: [params.fightID],
      reportID: params.reportID,
    },
    [PF_RED_BUFF.unit, PF_GREEN_BUFF.unit, PF_PURPLE_BUFF.unit]
  );

  const [redDeathEvents, greenDeathEvents, purpleDeathEvents] =
    await Promise.all<DeathEvent[]>(
      [red, green, purple].map(async (sourceID) => {
        if (!sourceID) {
          return [];
        }

        return getEvents<DeathEvent>({
          ...params,
          dataType: EventDataType.Deaths,
          hostilityType: HostilityType.Enemies,
          sourceID,
        });
      })
    );

  const earliestUnitDeathTimestamp = [
    ...redDeathEvents,
    ...greenDeathEvents,
    ...purpleDeathEvents,
  ].reduce(
    (acc, event) => (acc <= event.timestamp ? acc : event.timestamp),
    Infinity
  );

  const [redAuraApplication, greenAuraApplication, purpleAuraApplication] =
    await Promise.all(
      [PF_RED_BUFF.aura, PF_GREEN_BUFF.aura, PF_PURPLE_BUFF.aura].map(
        async (abilityID) => {
          if (earliestUnitDeathTimestamp === Infinity) {
            return [];
          }

          const events = await getEvents<ApplyBuffEvent | RemoveBuffEvent>({
            ...params,
            startTime: earliestUnitDeathTimestamp,
            dataType: EventDataType.Buffs,
            hostilityType: HostilityType.Friendlies,
            abilityID,
          });

          return events.filter(
            (event): event is ApplyBuffEvent => event.type === "applybuff"
          );
        }
      )
    );

  return [
    ...redDeathEvents,
    ...greenDeathEvents,
    ...purpleDeathEvents,
    ...redAuraApplication,
    ...greenAuraApplication,
    ...purpleAuraApplication,
  ];

  // return {
  //   red: {
  //     kills: redDeathEvents.length,
  //     consumed: redAuraApplication.length,
  //   },
  //   green: {
  //     kills: greenDeathEvents.length,
  //     consumed: greenAuraApplication.length,
  //   },
  //   purple: {
  //     kills: purpleDeathEvents.length,
  //     consumed: purpleAuraApplication.length,
  //   },
  // };
};
