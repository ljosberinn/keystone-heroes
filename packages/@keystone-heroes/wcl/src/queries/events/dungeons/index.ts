import { DungeonIDs } from "@keystone-heroes/db/data";
import type { PlayableClass } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  RemoveBuffEvent,
} from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  isDungeonWithEvent(id) ? dungeonExpressionMap[id] : [];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes,
  dungeonID: DungeonIDs,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
):
  | ApplyDebuffEvent[]
  | CastEvent[]
  | (DeathEvent | DamageEvent | ApplyBuffEvent)[]
  | (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  | (DamageEvent | HealEvent)[] => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, playerMetaInformation);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATRE_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    default:
      return [];
  }
};
