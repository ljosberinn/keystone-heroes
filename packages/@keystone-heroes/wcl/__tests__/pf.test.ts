import { PlayableClass } from "@keystone-heroes/db/types";
import { getPFEvents } from "@keystone-heroes/wcl/queries/events/dungeons/pf";

import allEvents from "./fixtures/allEvents.json";

describe("getPFEvents", () => {
  test("works", () => {
    expect(
      getPFEvents(allEvents, [
        {
          actorID: 62,
          class: PlayableClass.Priest,
        },
        {
          actorID: 1,
          class: PlayableClass.DemonHunter,
        },
        {
          actorID: 63,
          class: PlayableClass.Rogue,
        },
        {
          actorID: 65,
          class: PlayableClass.Rogue,
        },
        {
          actorID: 64,
          class: PlayableClass.Shaman,
        },
      ])
    ).toMatchSnapshot();
  });
});
