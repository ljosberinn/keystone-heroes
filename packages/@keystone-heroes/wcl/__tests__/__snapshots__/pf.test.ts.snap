// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPFEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 161,
    "targetInstance": 4,
    "timestamp": 9399646,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 161,
    "targetInstance": 5,
    "timestamp": 9400225,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 162,
    "targetInstance": 1,
    "timestamp": 9400225,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 161,
    "targetInstance": 6,
    "timestamp": 9430006,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 163,
    "targetInstance": 1,
    "timestamp": 9486481,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 163,
    "targetInstance": 2,
    "timestamp": 9769975,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 162,
    "targetInstance": 2,
    "timestamp": 9770061,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 163,
    "targetInstance": 3,
    "timestamp": 11090607,
    "type": "death",
  },
  Object {
    "abilityGameID": 0,
    "pin": "0",
    "sourceID": -1,
    "targetID": 161,
    "targetInstance": 7,
    "timestamp": 11099140,
    "type": "death",
  },
  Object {
    "abilityGameID": 340210,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4006933,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340210,
    "pin": "0",
    "sourceID": 63,
    "targetID": 63,
    "timestamp": 4076733,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340210,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4563925,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340210,
    "pin": "0",
    "sourceID": 63,
    "targetID": 63,
    "timestamp": 5260722,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340225,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4563925,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340225,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4979920,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340271,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4622916,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340271,
    "pin": "0",
    "sourceID": 65,
    "targetID": 65,
    "timestamp": 4978916,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 328501,
    "amount": 7781258,
    "hitType": 1,
    "pin": "0",
    "sourceID": 165,
    "sourceInstance": 4,
    "targetID": 160,
    "targetInstance": 2,
    "timestamp": 9074155,
    "type": "damage",
    "unmitigatedAmount": 70987,
  },
]
`;
